Add support of CentOS Stream 9

cathay4t opened the pull request:
Introducing new `dist_git_branches` -- `c9s` for CentOS stream 9
downstream build.

The `rebase_bug_no` is required for CentOS stream 9 dist-git commit
message.

To make it work, you need store gitlab token in `~/.config/packit.yml`,
like:

```config
authentication:
    gitlab.com:
        token: glpat-<secrets>
```

You need to run packit in the upstream git repo with command:

```bash
    packit --config ~/packit_ds_nispor.yml  propose-downstream
```

Example packit.yml(since nispor has no spec in upstream tarball, we use
/tmp/nispor.spec for test purpose):

```yml
---
downstream_package_name: nispor
upstream_project_url: https://github.com/nispor/nispor
issue_repository: https://github.com/nispor/nispor
update_release: false
upstream_package_name: nispor
upstream_tag_template: v{version}
copy_upstream_release_description: false

actions:
  post-upstream-clone:
    - bash -c "cp /tmp/nispor.spec ./"

jobs:
  - job: propose_downstream
    trigger: release
    rebase_bug_no: 2224363
    dist_git_branches:
      - c9s
```

cathay4t commented:
The centos stream dist-git PR created by packit of this PR is https://gitlab.com/redhat/centos-stream/rpms/nispor/-/merge_requests/13

lachmanfrantisek commented on packit/cli/propose_downstream.py:
```diff
@@ -11,20 +11,38 @@
 def get_dg_branches(api, dist_git_branch):
     cmdline_dg_branches = dist_git_branch.split(",") if dist_git_branch else []
+
+    # TODO(Gris): Are we support mixing c9s with fedora-all ?
```
I think it's fine to mix those, but:

* As I see, this will require improvements in how we work with `dist_git_base_url`.
* We can leave this to a follow-up.

cathay4t replied:
V2 patch support placing `c9s` with `fedora-all`.

lachmanfrantisek commented on packit/cli/propose_downstream.py:
```diff
@@ -140,6 +157,25 @@ def propose_downstream(
+        description = None
+        rebase_bug_no = None
+
+        # BUG(Gris Ge): buggy code on accessing jumble package_config data
```
I would move `rebase_bug_no` to the top level of config and don't cycle through the jobs when using this via CLI. WDYT?

In service, it's taken care of -- no cycling needed. (And you can define properties both globally and on the job level.)

cathay4t replied:
The `rebase_bug_no` is removed in V2 patch.

lachmanfrantisek commented on packit/config/common_package_config.py:
```diff
@@ -203,12 +209,23 @@ def __init__(
-        self.dist_git_namespace: str = dist_git_namespace or getenv(
-            "DISTGIT_NAMESPACE", DISTGIT_NAMESPACE
-        )
+        # TODO(Gris): Support mixing c9s and fedora-all branches
+        # TODO(Gris): Observed two PackageConfig been created with different
+        #             dist_git_branches, need investigation
+        if dist_git_branches and "c9s" in dist_git_branches:
+            self.dist_git_base_url: str = dist_git_base_url or getenv(
+                "DISTGIT_URL", CENTOS_STREAM_DISTGIT_URL
+            )
+            self.dist_git_namespace: str = dist_git_namespace or getenv(
+                "DISTGIT_NAMESPACE", CENTOS_STREAM_DISTGIT_NAMESPACE
+            )
+        else:
+            self.dist_git_base_url: str = dist_git_base_url or getenv(
+                "DISTGIT_URL", PROD_DISTGIT_URL
+            )
+            self.dist_git_namespace: str = dist_git_namespace or getenv(
+                "DISTGIT_NAMESPACE", DISTGIT_NAMESPACE
+            )
```
I am wondering if this shouldn't be moved to the job implementation so we can have `None` as the default that can be replaced when we process a specific branch.

cathay4t replied:
I have no better option. Any suggestion?

lachmanfrantisek reviewed (commented):
Very nice, Gris! Thank you!

Looking at the code, we should have done that years ago..;)

Let us know if we can help with anything!

lbarcziova commented on packit/cli/propose_downstream.py:
```diff
@@ -11,20 +11,38 @@
     cmdline_dg_branches = dist_git_branch.split(",") if dist_git_branch else []
+
+    # TODO(Gris): Are we support mixing c9s with fedora-all ?
+    if any(
+        job.type == JobType.propose_downstream
+        and any(
+            pkg_conf.dist_git_branches and "c9s" in pkg_conf.dist_git_branches
+            for pkg_conf in job.packages.values()
+        )
+        for job in api.package_config.jobs
+    ):
+        api.config.pkg_tool = "centpkg"
+        api.package_config.dist_git_base_url = CENTOS_STREAM_DISTGIT_URL
```
I think we should not touch jobs here (= in packit repo code) at all, as the jobs are service-specific...we should probably rather check the `dist_git_branch` argument, WDYT?

cathay4t replied:
In V2 patch, I created two instance of `PackitApi` as workaround. This require better design which I don't have resource to do.

lbarcziova commented on packit/cli/propose_downstream.py:
```diff
@@ -140,6 +157,25 @@ def propose_downstream(
+        rebase_bug_no = None
+
+        # BUG(Gris Ge): buggy code on accessing jumble package_config data
+        for job in api.package_config.jobs:
```
same as in my previous comment applies here

cathay4t replied:
Please check whether V2 patch is better.

lbarcziova reviewed (commented):
thanks for this work, Gris!

cathay4t commented:
V2 patch uploaded. Changes since V1:
 * Removed the `rebase_bug_no` option as CentOS stream gitlab allows commit has zero bug/jira reference during non-exception phase.
 * Create two instance of `PackitApi` to support using `c9s` and `fedora-all` branches in single packit.yml.

lachmanfrantisek reviewed (commented):
Thanks a lot, @cathay4t ! Without the Bugzilla IDs, the change can be much smaller.

cathay4t commented on packit/cli/propose_downstream.py:
```diff
@@ -64,6 +71,45 @@ def sync_release(
+            force,
+        )
+
+    # Check whether we still have Fedora branches to build
+    pending_fedora_works = False
+    for job in api.package_config.jobs:
+        if job.type == JobType.propose_downstream:
+            for pkg_conf in job.packages.values():
+                new_branches = []
+                for branch in pkg_conf.dist_git_branches:
+                    if branch != "c9s":
+                        new_branches.append(branch)
+                        pending_fedora_works = True
+                pkg_conf.dist_git_branches = new_branches
+
+    if not pending_fedora_works:
+        return
+
     branches_to_update = get_dg_branches(api, dist_git_branch)
```
The `get_dg_branches` will invoke `fedpkg`, which seems unnecessary when the YAML only has `c9s`.

cathay4t commented on packit/cli/propose_downstream.py:
```diff
@@ -230,3 +276,27 @@ def pull_from_upstream(
+    )
+
+
+def get_centos_stream_job(api):
```
How about the `api.config.pkg_tool`? Should I change it back to fedpkg when done?

mfocko commented:
Based on the discussion with @lachmanfrantisek, I'll reopen the PR from my branch with your changes and adjust where necessary (I'll also skim through the reviews done here). Thanks a lot :heart:
